### PR TITLE
Java Lab: Add JWT token generation for JavaBuilder

### PR DESCRIPTION
**Pull request by molly-moen**
This PR creates a controller that generates a JWT token for authorized users (either levelbuilders or participants in the CSA pilot), which will be used to authenticate with JavaBuilder. Java Lab requests the token, but doesn't do anything with it yet. The token contains the following information. The short names are the jwt standardized keys.

```
issued at time (iat)
expiration time (exp, set to 15 minutes from when token was issued)
session id (sid)
uid (user id)
storage_id
storage_app_id
channel_id
project_version
```

The JWT token is generated with an RSA key pair. The keys are stored via AWS, and there is one per environment.

If you want to test this locally, you will need to add the following lines to your locals.yml or development.yml (note if you add to locals it will cause unit tests to fail because it will try to load a secret from test). You can also generate your own private key and use that.

```
javabuilder_private_key: !Secret
javabuilder_key_password: !Secret
```

## Links

- [auth spec](https://docs.google.com/document/d/1edHZj1a_zZsLnTUqvmskSeGugNUprQVXt2bfWLdcDBM/edit#)
- [overall infrastructure spec](https://docs.google.com/document/d/196aKj947BYZXZH3nGvzHprgWPOoy2Kw1x3sKgziUaSo/edit#)
- [slack discussion on how we deal with keys](https://codedotorg.slack.com/archives/C03CK49G9/p1616712904170500)
- jira ticket: [CSA-183](https://codedotorg.atlassian.net/browse/CSA-183)

## Testing story
Added unit tests that verify only authorized users can generate a token, and that the token generates as expected. Also tested locally.

## Deployment strategy
All keys are in AWS Secrets Manager already, so this PR should not cause any issues due to the new key in `config.yml.erb`.

## Follow-up work
Create a Lambda authorizer that can decode this JWT token.

## Security
The private tokens are stored securely, and this feature is only enabled for authorized users. Currently getting a token does not enable you to do anything, but it will in the future.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/java_builder_sessions_controller.rb
Comment (molly-moen) [inline dashboard/app/controllers/java_builder_sessions_controller.rb]: should we include anything else here? One thing that comes to mind is user id, but I don't know if Java Builder will need it or not.

Comment (jmkulwik) [reply dashboard/app/controllers/java_builder_sessions_controller.rb]: JavaBuilder won't use the user id, but it will use the channel ID for the project. I don't think that needs to be included here, though, because the channel ID is public information.

**Review by jmkulwik — APPROVED** [review]
Looks good to me. I have to admit, my Ruby/Rails is pretty rusty - it may be worth getting a second opinion.

File: dashboard/config/routes.rb
Comment (jmkulwik) [inline dashboard/config/routes.rb]: Would it make sense to let this route automatically to `index` or `new`?

Comment (molly-moen) [reply dashboard/config/routes.rb]: good question--my concern with using one of those is it's not super clear what the function does in that case (get an access token).

**Comment by sureshc** [comment]
Probably worth defining what a JavaBuilder session is that we’re authorizing. A student could have multiple tabs open or the same JavaLab project open in two different browsers, if we have to support HTTP fallback when websocket is blocked by a school’s network, we may have trouble routing messages from the student’s browser(s) to the correct JavaBuilder long running Lambda.

Possible definition: A JavaBuilder session is a request from a signed in user asking JavaBuilder to:

1. Fetch a specific version of a specific project owned by the user
2. Compile that project version
3. Execute that project
4. Exchange messages between the user’s client and the executing project to exchange project input/output, interactive debugging, etc.

Given those possible requirements, I think it’s worth including the following in the token: user id, project id (storage_id and storage_app_id), channel_token (encrypted storage_id and storage_app_id), project version identifier, and a unique JavaBuilder session id generated by Code Studio for logging/tracking purposes.

**Comment by molly-moen** [comment]
@sureshc I added the fields you suggested to the token. Currently I'm just generating a session id and not logging it anywhere--should we add some sort of logging now or is it ok to leave as is until we need to log it?

**Comment by sureshc** [comment]
> should we add some sort of logging now or is it ok to leave as is until we need to log it?

I'm thinking it would be useful for the JavaLab client development environment to display the session id if something goes wrong and/or if a student clicks a button to open a Support issue, but it may be too cumbersome to log it in Code Studio. Maybe we could publish a Firehose event with all of the details in the token? Early in the pilot, it might be useful to have more data about how often sessions are being started.

I'm mostly thinking that we would have the synchronous Lambda and the long running Lambda include this session id in any entry they publish to debug logs (in CloudWatch). There are so many moving parts (API Gateway, synchronous Lambda, SQS, long running Lambda) that it would be handy to search in CloudWatch by a JavaBuilder session id and pull up all log entries related to that session to figure out what happened.

**Comment by molly-moen** [comment]
@sureshc I added a firehose log with the payload from the token and added the session id to the payload returned to the client, so we can do what you suggested and log the session id to the user if there's an issue.

**Comment by sureshc** [comment]
Sorry for the very belated suggestion here, but we may benefit from using the variable name filename prefix `javabuilder` instead of `java_builder` in Code Studio to make it easier for people searching our code base. This might provide some consistency with the javabuilder webservice codebase as well.

File: dashboard/app/controllers/javabuilder_sessions_controller.rb
Comment (sureshc) [inline dashboard/app/controllers/javabuilder_sessions_controller.rb:30]: Do JWT tokens have a length limit? If not, might be useful for people reading the JavaBuilder code to read non-abbreviated attributes in the token (`issued_at_time` instead of `iat`) and same for `exp`, `sid`, and `uid`

File: dashboard/app/models/ability.rb
Comment (sureshc) [inline dashboard/app/models/ability.rb:294]: I'm assuming that in the future we could grant this ability a bit more generically (any student in an AP CS A Section led by a verified teacher, or something similar).

File: dashboard/app/controllers/javabuilder_sessions_controller.rb
Comment (sureshc) [reply dashboard/app/controllers/javabuilder_sessions_controller.rb:30]: oops. I see these are JWT "reserved claim names". Nevermind!

File: dashboard/app/models/ability.rb
Comment (molly-moen) [reply dashboard/app/models/ability.rb:294]: yep! We can put whatever logic we want here.

**Review by sureshc — APPROVED** [review]
This is awesome!

**Pull request merged**
